Add ping task with fallback support and RTT logging

0x-fran opened the pull request:
This PR introduces a new `ping_task` which periodically (once a minute) pings a primary Stratum server and falls back to a secondary server if the primary is unreachable. It uses the ESP-IDF `ping_sock` to perform ICMP echo requests and logs the average round-trip time (RTT) if any replies are received.

**Changes include:**
- New source file `ping_task.cpp` in `tasks/`
- Declaration of `ping_task()` in `ping_task.h`
- Uses `perform_ping()` internally to handle the logic and RTT logging
- Task is launched via `xTaskCreate()` in `main.cpp`
- Update to CMakeLists.txt

**Purpose:**
Improves connection monitoring and reliability by enabling runtime checks of network reachability.

---

Let me know if anything should be rephrased or extended.

0x-fran commented:
Just cleaned up commit metadata (name/email). No code changes.

shufps commented:
nice! One thing, maybe we should do the ping for the currently active pool? It can be determined by:

```
STRATUM_MANAGER.isUsingFallback();
STRATUM_MANAGER.isAnyConnected();
```

duckAxe commented:
Extending the `/api/system/info` endpoint to allow WebGUI to output the RTT value would be great.

0x-fran commented:
Added `ping_task` RTT value to system UI and `api/system/info`
- New field `lastpingrtt` passed from ping_task to frontend
- Displayed in UI with 2 decimal precision

0x-fran commented:
Fix: support DNS resolution for ping targets

Replaced `inet_aton() `with `getaddrinfo()` to support hostnames in `perform_ping()`

0x-fran commented:
**Refactor**: Extract DNS resolution to `dns_task` module with caching and TTL

- Moved DNS resolution logic from `ping_task` to new `dns_task.[cpp|h]`
- Added TTL-based DNS caching to reduce redundant lookups
- Supports both FQDN and direct IP usage transparently
- Updated `ping_task` to use `dns_task::resolve_hostname_to_ip()`
- Updated `CMakeLists.txt` to include `dns_task`
- `dns_task` is now reusable across the codebase, e.g. in `stratum_task.cpp`

0x-fran commented:
Cleaned up the commit history – removed unrelated `Feature: Filter system logs by input on WebUI` changes.
Only DNS and ping-related changes remain.
Thanks for your patience! Let me know if anything else should be adjusted

duckAxe commented on main/http_server/axe-os/src/app/pages/home/home.component.html:
```diff
@@ -266,6 +266,16 @@ <h4>Loading...</h4>
+                  <span class="text-hint">Ping:</span>
+                </div>
+                <div class="col">
+		  {{ info.lastpingrtt !== undefined ? info.lastpingrtt.toFixed(2) + ' ms (avg)' : 'n/a' }}
```
Please indent code line

0x-fran commented:
This change reduces the memory footprint of the ping task by:

- Moving RTT statistics struct to static memory
- Reusing a static `esp_ping_callbacks_t` instance
- avoiding heap allocation

These adjustments help prevent memory exhaustion issues (e.g., only 28 bytes remaining) observed during long runtimes causing software panic reset. Access to dns_cache in `resolve_hostname()` is now synchronized via `std::mutex` to prevent race conditions when accessed concurrently.

shufps commented on main/tasks/dns_task.cpp:
```diff
@@ -0,0 +1,89 @@
+    time_t timestamp;
+};
+
+static std::map<std::string, CachedEntry> dns_cache;
```
Hmm I'm not so keen on using such complet dynamic memory things from the STL. Imho we don't need a STL map for just cashing a single string :sweat_smile: 

In overall the functions look a lot more complicated than it need to be

0x-fran commented:
## Finalized PR Status: Simplified, Unified, and Ready for Review 🚀

Thank you for all the feedback and notes during this process! What started as a straightforward `ping_task` addition evolved into a broader stabilization of DNS and RTT handling. This PR is now focused on reducing complexity, improving memory usage, and ensuring long-term maintainability of shared logic across tasks.

### Summary of Final Changes
1. **Simplified and unified DNS handling:**
   - Created `dns_task` as a centralized DNS resolver to replace separate implementations in `ping_task` and `stratum_task`.
   - Replaced `std::map`-based DNS caching with a static 2-slot array supporting TTL-based expiration – simpler, leaner, and ideal for embedded systems.
   - Resolved race conditions by synchronizing DNS cache access with FreeRTOS mutexes (`SemaphoreHandle_t`).

2. **Improved RTT logic and reduced memory footprint:**
   - Fully removed dynamic memory usage in DNS cache and RTT statistics (`dns_task` and `ping_task`) by relying on static and preallocated structures.
   - Optimized RTT stats memory (from 12B → 6B) and reset logic on hostname changes or TTL expiry.
   - System-level calls like `getaddrinfo` (dns) and `esp_ping_new_session` (ping) now have proper handling and clean deallocation, ensuring they do not leak.

3. **General refactoring and code cleanup:**
   - Merged the DNS resolution logic in `dns_task`, which is now shared across `ping_task` and `stratum_task`. This reduces duplicated logic and adds caching for improved performance.
   - Improved RTT handling in `ping_task` with more robust overflow guards and refined logging for long-running scenarios.
   - Added `dns_task_init()` in `main.cpp` for clean task initialization during startup.

### Rationale for Changes
The primary focus has been to address feedback around STL/dynamic memory use while introducing long-term optimizations:
- Embedded-safe logic with minimized dynamic memory usage.
- Unified and shared DNS and RTT handling across tasks.
- A substantial reduction in complexity and maintainability overhead.

### Commit Cleanup
There are 27 commits in total, nearly half resulting from the iterative feedback process and structural changes.
I’m happy to squash/rebase them into a cleaner history if needed – just let me know!

---

Big thanks to everyone for your guidance throughout this PR! It’s been a valuable learning process, and I’m excited to hear your thoughts on the revised implementation 😊

shufps commented:
Thx! I'll test it shortly :blush: :+1:

shufps commented:
Thx for revising :raised_hands:  Ah I leave it up to you to, I'm a pragmatic guy and the end result is what counts :sweat_smile:

0x-fran commented:
Time to let this glorious mess rest in peace 😅  
As promised: closing this beautifully chaotic first attempt in favor of #175 ... a cleaner, leaner version built on `develop`, without DNS detours, record-breaking RTT paranoia, or speculative DNS caching in case the IP changes during a solar flare 😇

Thanks again for the patience and trust 🙌
